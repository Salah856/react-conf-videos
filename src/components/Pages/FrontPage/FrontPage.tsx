import * as React from 'react'
import { connect } from 'react-redux'
import { compose, pure, withHandlers } from 'recompose'

import {
  List,
  Meta,
  Header,
  SearchInput,
  ResultDetails,
  InnerLayoutContainer
} from 'components'
import { IndexedConferences } from '../../../domain'
import { routingActions, ApplicationState } from 'redux/modules'

type ReduxProps = {
  conferences: IndexedConferences,
  filterValue: string
}

type DispatchpProps = {
  navigateToSearchURL: typeof routingActions.navigateToSearchURL
}

type WithHandlers = {
  onInputChange: (e: React.ChangeEvent<HTMLInputElement>) => void
}

type Props = ReduxProps & DispatchpProps & WithHandlers

export const FrontPageInner: React.SFC<Props> = (props) => (
  <div>
    <Meta title={props.filterValue} />
    <Header />
    <InnerLayoutContainer>
      <SearchInput
        filterValue={props.filterValue}
        onChange={props.onInputChange}
      />
      <ResultDetails conferences={props.conferences} />
      <List conferences={props.conferences} />
    </InnerLayoutContainer>
  </div>
)

const mapStateToProps = ({search: {conferences, filterValue}}: ApplicationState) => ({
  filterValue,
  conferences
})

const dispatchMap = ({
  navigateToSearchURL: routingActions.navigateToSearchURL
})

export const FrontPage = compose<Props, {}>(
  connect(mapStateToProps, dispatchMap),
  pure,
  withHandlers<ReduxProps & DispatchpProps, WithHandlers>({
    onInputChange: ({navigateToSearchURL}) => (e: React.ChangeEvent<HTMLInputElement>) => {
      navigateToSearchURL(e.target.value)
    }
  })
)(FrontPageInner)
