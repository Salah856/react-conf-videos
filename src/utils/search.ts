import { compose, toLower, trim } from 'ramda'
import { remove as removeDiacritics } from 'diacritics'

export const cleanQuery = compose(
  removeDiacritics,
  toLower,
  trim
)

export const isFilterEmpty = (filterValue: string = '') => filterValue.trim() === ''
