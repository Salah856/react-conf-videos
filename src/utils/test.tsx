import { mount, shallow } from 'enzyme';
import * as React from 'react';
import { ReactElement } from 'react';
import { MemoryRouter } from 'react-router';
import { StateObservable } from 'redux-observable';
import { createMockStore } from 'redux-test-utils';
import { Subject } from 'rxjs';
import * as Immutable from 'seamless-immutable';

import { ApplicationState } from 'redux/modules';
import {
  Conference,
  IndexedConferences,
  IndexedPresenters,
  IndexedVideos,
  Presenter,
  Video
} from '../domain';

// cast to ApplicationState so TS doesn't complain when sending in partials to Epics for testing
export const mockStore = (state = {}) =>
  new StateObservable<ApplicationState>(
    new Subject(),
    state as ApplicationState
  );

const isStore = (store: any) => store.hasOwnProperty('isActionTypeDispatched');

// shallow mount component with a store in context
export const shallowWithStore = (storeP: any, component: ReactElement<{}>) => {
  const store = isStore(storeP) ? storeP : createMockStore(storeP);
  return shallow<any, any>(component, { context: { store } });
};

export const mountWithStore = (storeP: any, component: ReactElement<{}>) => {
  const store = isStore(storeP) ? storeP : createMockStore(storeP);
  return mount<any, any>(component, { context: { store } });
};

export const wrapWithMemoryRouter = (Component: React.ReactElement<any>) => (
  // key within location prop keeps changing, so manually set
  // so snapshorts are consistent
  <MemoryRouter initialEntries={[{ key: 'xxxx' }]}>{Component}</MemoryRouter>
);

/** Mock data */
export const mockVideo = (title = 'test title'): Video => ({
  link: 'a link',
  split: '',
  title: title || 'test title',
  length: '12:34',
  lightning: false,
  presenter: 'siomn carter',
  embeddableLink: 'a link'
});

export const mockVideos = (): IndexedVideos => ({
  aaa: mockVideo('aaa'),
  bbb: mockVideo('bbb'),
  ccc: mockVideo('ccc')
});

export const mockPresenter = (name = 'simon carter'): Presenter => ({
  name
});

export const mockPresenters = (): IndexedPresenters => ({
  xxx: mockPresenter(),
  zzz: mockPresenter('zzz')
});

export const mockConference = (): Conference => ({
  date: 'XX/YY/ZZZ',
  title: 'fake conf title',
  website: 'fake url',
  playlist: 'day 1',
  videos: ['aaa', 'bbb', 'ccc']
});

export const mockIndexedConferences = (): IndexedConferences => ({
  ['XXXX']: mockConference()
});

export const mockRouterState = (
  pathname = '/conference/8424e37df85b9eccbe48e9a55d93845e/react-conf-2018'
) => ({
  router: {
    location: {
      pathname,
      search: '',
      hash: ''
    }
  }
});

export const mockConferencePageSlice = () => ({
  conferencePage: {
    selectedConferenceId: 'XXXX',
    conference: mockConference()
  }
});

export const mockDataState = () =>
  Immutable({
    data: {
      conferences: mockIndexedConferences(),
      presenters: mockPresenters(),
      videos: mockVideos()
    }
  });
