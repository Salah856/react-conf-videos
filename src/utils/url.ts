import { compose, replace, toLower } from 'ramda';

// replaces white space with - and lower case
export const sluggifyUrl = compose(
  replace(/\s/g, '-'),
  toLower
);
