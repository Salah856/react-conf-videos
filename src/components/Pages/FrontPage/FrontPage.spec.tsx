import * as React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import { shallowWithStore, mockConference } from 'utils'
import FrontPage, { FrontPageInner } from './FrontPage'

describe('FrontPage', () => {
  describe('FrontPageInner', () => {

    const shallowComp = (isActive: boolean) => shallow(
      <FrontPageInner
        conferences={{x: mockConference()}}
        filterValue=""
        navigateToSearchURL={jest.fn()}
        onInputChange={() => null}
      />
    )

    it('should render', () => {
      // arrange
      // act
      const wrapper = shallowComp(true)

      // expect
      expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('should not have the isActive classes if isActive is false', () => {
      // arrange
      // act
      const wrapper = shallowComp(false)

      // assert
      expect(wrapper.find('.frontPage.isActive').length).toBe(0)
      expect(wrapper.find('.text.isActive').length).toBe(0)
    })
  })

  describe('FrontPage', () => {
    it('should render', () => {
      // arrange
      const store = {
        frontPage: {isActive: false},
        search: {
          conferences: {X: mockConference()},
          filterValue: ''
        }
      }

      // act
      const wrapper = shallowWithStore(store, <FrontPage />)

      // assert
      expect(toJson(wrapper)).toMatchSnapshot()
    })
  })
})
