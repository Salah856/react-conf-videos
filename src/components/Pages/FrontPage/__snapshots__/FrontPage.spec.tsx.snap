// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FrontPage FrontPageInner should render 1`] = `
<Fragment>
  <Component
    title=""
  />
  <Component
    tagline="Search React.js conference videos."
    title="React.js Videos"
    titleLink="/#/search"
  />
  <Component>
    <Component
      filterValue=""
      onChange={[Function]}
    />
    <Component
      conferences={
        Object {
          "x": Object {
            "date": "XX/YY/ZZZ",
            "playlist": "day 1",
            "title": "react conf 2018",
            "videos": Array [
              "aaa",
              "bbb",
              "ccc",
            ],
            "website": "fake url",
          },
        }
      }
    />
    <Component
      conferences={
        Object {
          "x": Object {
            "date": "XX/YY/ZZZ",
            "playlist": "day 1",
            "title": "react conf 2018",
            "videos": Array [
              "aaa",
              "bbb",
              "ccc",
            ],
            "website": "fake url",
          },
        }
      }
    />
  </Component>
</Fragment>
`;
